### ISSUE TITLE
[Bug]: mutation_aspect for FancyBboxPatch does wrong scaling

### ISSUE THREAD

**Issue by Balthasar-eu** [matplotlib/matplotlib#28876]
### Bug summary

The scaling when using mutation_aspect is wrong. I attached a demonstration picture that uses code from the [FancyBboxPatch demo](https://matplotlib.org/stable/gallery/shapes_and_collections/fancybox_demo.html#sphx-glr-gallery-shapes-and-collections-fancybox-demo-py).

I have a fix and opened a pull request.

### Code for reproduction

```Python

import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import matplotlib.transforms as mtransforms
from matplotlib.path import Path
from importlib import reload
import math

reload(mpatches)

ASPECT = 2
PAD = 0.2
ASQRT = math.sqrt(ASPECT)
BSTYLE = f"round,pad={PAD}"

APPLY_FIX = False

if APPLY_FIX:

    def get_path(self):
        """Return the mutated path of the rectangle."""
        boxstyle = self.get_boxstyle()
        m_aspect = math.sqrt(self.get_mutation_aspect())
        # Call boxstyle with y, height squeezed by aspect_ratio.
        path = boxstyle(
            self._x * m_aspect,
            self._y / m_aspect,
            self._width * m_aspect,
            self._height / m_aspect,
            self.get_mutation_scale(),
        )
        return Path(
            path.vertices * [1 / m_aspect, m_aspect],
            path.codes,
        )

    mpatches.FancyBboxPatch.get_path = get_path


def add_fancy_patch_around(ax, bb, **kwargs):
    fancy = mpatches.FancyBboxPatch(
        bb.p0,
        bb.width,
        bb.height,
        fc=(1, 0.8, 1, 0.5),
        ec=(1, 0.5, 1, 0.5),
        **kwargs,
    )
    ax.add_patch(fancy)
    return fancy


def draw_control_points_for_patches(ax):
    for patch in ax.patches:
        patch.axes.plot(
            *patch.get_path().vertices.T, ".", c=patch.get_edgecolor()
        )


fig, axs = plt.subplots(2, 2, figsize=(8, 8))

bb = mtransforms.Bbox(
    [[-0.1 * ASQRT, -0.1 / ASQRT], [0.1 * ASQRT, 0.1 / ASQRT]]
)

ax = axs[0, 0]
fancy = add_fancy_patch_around(ax, bb, boxstyle=BSTYLE)
ax.set(
    xlim=(-0.5, 0.5),
    ylim=(-0.5, 0.5),
    aspect=1,
    title=f'boxstyle="{BSTYLE}"\nmutation_aspect=1,aspect=1',
)
draw_control_points_for_patches(ax)
fancy = add_fancy_patch_around(ax, bb, boxstyle="square,pad=0")
fancy.set(edgecolor="black", facecolor="none", zorder=10)

bb = mtransforms.Bbox(
    [[-0.1 / ASQRT, -0.1 * ASQRT], [0.1 / ASQRT, 0.1 * ASQRT]]
)

ax = axs[0, 1]
fancy = add_fancy_patch_around(ax, bb, boxstyle=BSTYLE)
ax.set(
    xlim=(-0.5, 0.5),
    ylim=(-0.5, 0.5),
    aspect=1,
    title=f'boxstyle="{BSTYLE}"\nmutation_aspect=1,aspect=1',
)
draw_control_points_for_patches(ax)
fancy = add_fancy_patch_around(ax, bb, boxstyle="square,pad=0")
fancy.set(edgecolor="black", facecolor="none", zorder=10)

### BOTTOM ROW, aspect != 1

bb = mtransforms.Bbox([[-0.1, -0.1], [0.1, 0.1]])

ax = axs[1, 0]

fancy = add_fancy_patch_around(ax, bb, boxstyle=BSTYLE)
fancy.set(facecolor="none", edgecolor="green")
fancy = add_fancy_patch_around(ax, bb, boxstyle=BSTYLE, mutation_aspect=ASPECT)


ax.set(
    xlim=(-0.5 / ASQRT, 0.5 / ASQRT),  # (-0.5, 1.5),
    ylim=(-0.5 * ASQRT, 0.5 * ASQRT),  # (0, 1),
    aspect=1 / ASPECT,
    title=f'boxstyle="{BSTYLE}"\nmutation_aspect={ASPECT},aspect={1/ASPECT}',
)

draw_control_points_for_patches(ax)
fancy = add_fancy_patch_around(ax, bb, boxstyle="square,pad=0")
fancy.set(edgecolor="black", facecolor="none", zorder=10)

ax = axs[1, 1]

fancy = add_fancy_patch_around(ax, bb, boxstyle=BSTYLE)
fancy.set(facecolor="none", edgecolor="green")
fancy = add_fancy_patch_around(
    ax, bb, boxstyle=BSTYLE, mutation_aspect=1 / ASPECT
)

ax.set(
    xlim=(-0.5 * ASQRT, 0.5 * ASQRT),
    ylim=(-0.5 / ASQRT, 0.5 / ASQRT),
    aspect=ASPECT,
    title=f'boxstyle="{BSTYLE}"\nmutation_aspect={1/ASPECT},aspect={ASPECT}',
)

draw_control_points_for_patches(ax)
fancy = add_fancy_patch_around(ax, bb, boxstyle="square,pad=0")
fancy.set(edgecolor="black", facecolor="none", zorder=10)

fig.tight_layout()

plt.show()
```

### Actual outcome

![matplotlib_mutation_aspect_fixed](https://github.com/user-attachments/assets/78f11f87-948d-4268-a7c2-35d011b44fff)

### Expected outcome

![matplotlib_mutation_aspect](https://github.com/user-attachments/assets/3a4edd72-1d47-4e80-967b-949919e523ac)

### Additional information

Currently the Y-axis is scaled wrongly and the X-axis is not scaled at all. This is masked in the [demonstration](https://matplotlib.org/stable/gallery/shapes_and_collections/fancybox_demo.html#sphx-glr-gallery-shapes-and-collections-fancybox-demo-py) (last plot on the bottom of the page), because the non-scaled bbox in green has a different padding than the scaled one, so it seems like the scaling is correct there. The plots above demonstrate the issue with an aspect ratio of 4 and 0.25 and corresponding inverted mutation_scale for the fancy box, as well as the result using the code in the linked pull request. In the top row are plots that have the same visual scaling as, but with an aspect ratio of 1.

### Operating system

Fedora

### Matplotlib Version

3.8.2

### Matplotlib Backend

module://matplotlib_inline.backend_inline

### Python version

3.12.3

### Installation

pip

**Comment by Balthasar-eu** [comment]
I updated the plots to use an aspect of 2, so the points would not be outside the plot boundaries. You can copy the code and on the top change the parameters to see the problem in different aspect ratios and with the fix applied or not.

**Comment by timhoffm** [comment]
Thanks for the report. I'm not clear that this is a bug. The [documentation says](https://matplotlib.org/stable/api/_as_gen/matplotlib.patches.FancyBboxPatch.html):

```
mutation_aspect : float, default: 1
    The height of the rectangle will be squeezed by this value before the mutation and the
    mutated box will be stretched by the inverse of it. For example, this allows different
    horizontal and vertical padding.
```
While it's not 100% explicit, I read this as mutation scale only affects the vertical of the box.
It may be slightly more clear with "and the *height of* the mutated box will be stretched".
But essentially, it's explicitly stated that squeesing only affects the hight, and stretching
must be the inverse. I believe this is what is implemented.

**Comment by Balthasar-eu** [comment]
Thanks for the reply. I actually saw that in the description of the FancyBbox when I looked more into it. However I wonder how useful it is to keep the mutation_aspect how it is currently.

Intuitively I would assume mutation_aspect corrects for the inverse aspect. So 1/4 corrects for aspect 4. Indeed it does in some way, as it keeps the ratio between X and Y equal. However, this is done in a non-symmetric way and compensates aspects > 1 by shrinking the patch and compensating aspects < 1 by enlargening the patch.

Here is a demonstration for aspect 100 and 0.01:

![aspect100](https://github.com/user-attachments/assets/9d81a244-564e-49dd-9cfa-30d43bfd1cb8)

This basically only makes this useful if you want to keep the X-axis coordinates but not Y for some reason. A FancyBbox does not let you control outer coordinates anyways though, so I doubt the usefulness of this. (Which, on a side note, I would actually like as an option similar to margin and padding in html and I would also be happy to make a separate issue and also create a pull request. In the past I implemented my own box styles for that.). And again the break of symmetry here means that if x and y axis are exchanged in a plot the figures will look radically different. So while this might not necessarily be seen as a bug, as it is currently documented according to what the parameter does. I think that in most cases the wanted use of mutation scale is to keep the visual size of the plot stable, which is only achieved by scaling X and Y by the square root of the aspect. So the only way to achieve that currently is to also change either the padding or the coordinates.

I think a compromise would be to allow X and Y scaling but also have optimal visual scaling (√aspect) as an option.

> is a questionable statement. The box parameters are in box coordinates, i.e. it adds a pad of 0.3 to all sides. So the green is expected. The confusion arises from mixture of axes aspect and box aspect. It would be better to leave the Axes aspect as in the other examples and simply illustrate the docstring "For example, this allows different horizontal and vertical padding."

This is actually not true, as the green box in the plot has a 0.2 padding, when it should be 0.3:

![Screenshot_20240927_195527](https://github.com/user-attachments/assets/fd893d2e-36bc-4807-a540-fe2a43407ab2)

**Comment by timhoffm** [comment]
As the documentation states, first and foremost `mutation_aspect` allows different horizontal and vertical padding. This doesn't necessarily have anything to do with aspect. Say, you want a horizontal pad of 3 and a vertical pad of 1.5, then you can use `pad=3, mutation_scale=0.5`. Agreed the approach is not ideal, but with your suggestion you'd have to calculate an artifical pad, I suppose `pad=np.sqrt(3**2 + 1.5**), mutation_scale=0.3`. This is making this use case harder.

For the aspect use case, I also have doubts how widely used it is. It only works (relatively easily) if you manually set an Axes aspect. Otherwise, you don't have the necessary aspect info.

I've rewritten part of the example in #28900:
https://output.circle-artifacts.com/output/job/653663c2-9e74-4106-a7a9-4467336484ee/artifacts/0/doc/build/html/gallery/shapes_and_collections/fancybox_demo.html#parameters-for-modifying-the-box
Does this make it more clear?



`mutation_aspect` works according to its definition and one can achieve desired effects when giving the right number. I therefore wouldn't count it as a bug. Since we have high bars on breaking compatibility I don't think we can or should modify its behavior.

Overall, I have serious doubts on the `mutation_aspect` parameter. It's a bad API as it's not a good parametrization for any of the discussed use cases - and also your proposed change doesn't make it really better.

> Which, on a side note, I would actually like as an option similar to margin and padding in html and I would also be happy to make a separate issue and also create a pull request. In the past I implemented my own box styles for that.

I'm open for a discussion on other functionality, but then let's do the design before starting implementation of concrete aspects.

**Comment by Balthasar-eu** [comment]
> For the aspect use case, I also have doubts how widely used it is. It only works (relatively easily) if you manually set an Axes aspect. Otherwise, you don't have the necessary aspect info.

True, but that is the same for the situation right now. You should be able to just use the output of ax.get_aspect() though, no?

Anyways, I think the documentation is better, but I would maybe add that for a padding that is visually equivalent to a box with the same edge ratio in aspect 1 not just in relation between x- and y-padding but x- and y-scale of the overall box the padding size needs to be adjusted as well by the square root of the mutation_aspect. Something like this:

```Python
boxstyle=f"round,pad={pad / math.sqrt(m_aspect)}", mutation_aspect=m_aspect
```

I spend quite some time being frustrated at mutation_aspect until I figured this out. I don't think many people use these classes directly, but that should hopefully be helpful for someone else.

**Comment by timhoffm** [comment]
`ax.get_aspect()` returns "auto" unless the user explicitly defines an aspect, so that doesn't help. - Well you can calculate the actual aspect somehow, but I think there's no easy function, and don't know how to do it off the top of my head.

More importantly, the axes aspect is only relevant if you explicitly create the `FancyBboxPatch` and add it to the Axes. I believe this is comparably rare. The most common use case is drawing boxes around text, i.e. the use through `ax.text(..., bbox=dict(boxstyle=...))`. However there, the coordinate system is independent of the axes coordinates. The "unit" is text height and any pad is added visually equal to all sided. We don't need `mutation_aspect` there for corrections. It's only use in this context is being able to specify different horizontal and vertical pad.

But apparently, you have a case for drawing fancy boxes directly. Could you tell me what kind of visualization you want to achieve with this?

**Comment by Balthasar-eu** [comment]
I made an ideogram package using matplotlib: https://github.com/Balthasar-eu/pyideogram
At the core I a stacked barchart for the ideogram, but I replace the centromer and telomer segments using FancyBbox with a custom side round box style. The gene tracks are also all FancyBbox. I use custom style though that don't use padding and instead have their mutation inside the box, because I need to control their exact coordinates.

The aspect ratio of these can be in the area of 10,000,000, because I use base pair coordinates vs a single feature plotting axis. I solved the issue mostly internally, but I was never able to get it 100% correct or use mutation_aspect in any way (Although I don't use FancyBbox as intended (probably), so I did look into it too deeply back then). Recently, however, I used a FancyBbox in an unrelated plot and actually spend the time investigating what mutation_aspect actually does and found that it does not keep visual properties equal. I found out how to circumvent that now optimally, but it took me being frustrated the second time about this.

And looking back I actually use something more complicated than get_aspect there to get the ratio, although I copied that from Stackoverflow:
https://github.com/Balthasar-eu/pyideogram/blob/master/src/pyideogram/ideogram.py#L206

**Comment by timhoffm** [comment]
Ok, I count this as as an exotic use case for FancyBboxPatch 😄. 

Have you considered using PathCollection instead of the fancy boxes? Paths can contain Bezier segments, which should make it quite easy to create the desired effect.

Given that documentation and implementation align and that `mutation_scale` can be used (with maybe more numeric acrobatics in rarer cases), I don't see a case for a breaking behavior change. The doc update #28900 should be sufficient. You are welcome to comment there if you think something is missing.

**Comment by Balthasar-eu** [comment]
Sorry I forgot to reply, but thanks for taking the time and improving the documentation!
